src/Client/Speedrun.java

```diff
@@ -0,0 +1,329 @@
+            Client.displayMessage("@red@Exited from loadSpeedrun() in a strange way. Not starting speedrun.", Client.CHAT_QUEST);
+            break;
+        }
+      }
+    }
+  }
+
+  private static void beginSpeedrun(long timeCalled) {
+    active = true;
+    addStartTime(timeCalled);
+  }
+
+  private static void resetSpeedrun() {
+    startTimes.clear();
+    endTimes.clear();
+    completionTicks = new int[goalsDefined];
+    completionTimes = new long[goalsDefined];
+  }
+
+  public static void addStartTime(long startTime) {
+    startTimes.add(startTime);
+    Client.displayMessage(String.format("Starting segment at millisecond %d!!", startTime), Client.CHAT_QUEST);
+  }
+
+  public static void endTheRun() {
+    Settings.SPEEDRUNNER_MODE_ACTIVE.put(Settings.currentProfile, false);
+    endTheRUNNN = true;
+    saveAndQuitSpeedrun();
+  }
+
+  public static void addEndTime(long endTime) {
+    endTimes.add(endTime);
+    Client.displayMessage(String.format("Ending segment at millisecond %d!!", endTime), Client.CHAT_QUEST);
+    printTimeSinceLastSegment(true);
+    printTotalTicks();
+  }
+
+  public static void printGoalCompletion(String name, int ticks, long time) {
+    Client.displayMessage(String.format("Completed %s in %d ticks! (%d millis since last segment)",
+        name, ticks, calcTimeSinceLastSegment(time)), Client.CHAT_QUEST);
+  }
+
+  public static void printTimeSinceLastSegment(boolean referenceEndTimes) {
+    long timeDelta;
+    if (referenceEndTimes) {
+      timeDelta = endTimes.get(endTimes.size() - 1) - startTimes.get(startTimes.size() - 1);
+    } else {
+      timeDelta = calcTimeSinceLastSegment(System.currentTimeMillis());
+    }
+    Client.displayMessage(String.format("Millis since last time segment: %d", timeDelta), Client.CHAT_QUEST);
+  }
+
+  public static long calcTimeSinceLastSegment(long millis) {
+    int startSize = startTimes.size();
+    if (startSize >= 1)
+      return millis - startTimes.get(startSize - 1);
+    else
+      return millis;
+  }
+
+  public static void incrementTicks() {
+    if (active)
+      totalTicks++;
+  }
+
+  public static void printTotalTicks() {
+    Client.displayMessage(String.format("Total ticks spent in speedrun: %d", totalTicks), Client.CHAT_QUEST);
+  }
+
+  public static void saveAndQuitSpeedrun() {
+    loadResult = REASON_NEED_LOAD;
+    if (active) {
+      active = false;
+      addEndTime(System.currentTimeMillis());
+      try {
+        DataOutputStream speedrunData = new DataOutputStream(
+          new BufferedOutputStream(
+              new FileOutputStream(
+                  new File(
+                  Settings.Dir.SPEEDRUN + String.format(
+                      "/%d.%s.bin", startTimes.get(0), Client.username_login.replaceAll("[^a-zA-Z0-9]", ""))
+                  )
+              )
+          )
+        );
+
+        // Header information
+        speedrunData.writeInt(1533546026); // unix timestamp that RSC was taken offline
```

Comment:
probably good idea to add human like representation of that date in comments

Reply:
resolved